### PR TITLE
Add categories logic

### PR DESCRIPTION
**Pull request by artemklishch**
BREAKING CHANGES:

*Made extra liquibase files to create categories table, insert data, create books_categories table
 *Mase respective model, mapper, controller, service and interface
 *Updated books logic from the point of the categories

Task: https://mate.academy/learn/java-spring-boot/java-spring-boot-practice-v2/linear?learnItemsFilter=All&section=practice&testTaskSlug=java_add_category_model

### CODE REVIEW

File: src/main/java/org/example/intro/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+@Tag(name = "Category management", description = "Endpoints for categories management")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CategoryController.java]: or
```suggestion
    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
```

File: src/main/java/org/example/intro/dto/book/CreateBookDto.java [hunk 2]
```diff
@@ -24,4 +27,6 @@ public class CreateBookDto {
     String isbn;
     @Size(min = 10)
     String coverImage;
+    @NotBlankList(message = "Book has to have at least one category")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/dto/book/CreateBookDto.java]:
```suggestion
    @NotEmpty(message = "Book has to have at least one category")
```

for collections

File: src/main/java/org/example/intro/model/Book.java [hunk 3]
```diff
@@ -35,4 +43,14 @@ public class Book {
     private String coverImage;
     @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/org/example/intro/model/Book.java [hunk 4]
```diff
@@ -35,4 +43,14 @@ public class Book {
...
+    @JoinTable(
+            name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id")
+    )
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/Book.java]:
```suggestion
```

File: src/main/java/org/example/intro/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@NoArgsConstructor
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    private String name;
+    private String description;
+    @Column(name = "is_deleted", nullable = false)
+    private Boolean isDeleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/Category.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/org/example/intro/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -24,20 +28,19 @@ public class BookServiceImpl implements BookService {
...
         return bookRepository.findAll(pageable).stream()
-                .map(bookMapper::toBookDto)
-                .toList();
+                .map(bookMapper::toDto).toList();
     }
 
     @Override
     public BookDto findById(Long id) {
-        return bookMapper.toBookDto(bookRepository.findById(id).orElse(null));
+        return bookMapper.toDto(bookRepository.findById(id).orElse(null));
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/BookServiceImpl.java]: don't use orElse(null));, use orElseThrow instead

File: src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream()
+                .map(categoryMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        return categoryMapper.toDto(
+                categoryRepository.findById(id).orElse(null)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java]: same

File: src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public CategoryDto update(Long id, CreateCategoryRequestDto categoryDto) {
+        Category categoryFromDb = categoryRepository.findById(id).orElse(null);
+        if (categoryFromDb == null) {
+            throw new NoSuchElementException("Category with the id " + id + " does not exist");
+        }
+        categoryFromDb.setName(categoryDto.getName());
+        categoryFromDb.setDescription(categoryDto.getDescription());
+        categoryRepository.save(categoryFromDb);
+        return categoryMapper.toDto(categoryFromDb);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java]:
```suggestion
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(
                        "Category with the id " + id + " does not exist"
                ));
        categoryMapper.updateCategoryFromDto(categoryDto, category);
        return categoryMapper.toDto(repository.save(category));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java (hunk 8 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java]: ++, thank you so much. updated

File: src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java (hunk 7 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java]: ++, updated.

File: src/main/java/org/example/intro/service/impl/BookServiceImpl.java (hunk 6 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/BookServiceImpl.java]: ++, updated

File: src/main/java/org/example/intro/model/Category.java (hunk 5 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/Category.java]: ++, thank you. updated.

File: src/main/java/org/example/intro/model/Book.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/Book.java]: ++, Thank you. Updated.

File: src/main/java/org/example/intro/dto/book/CreateBookDto.java (hunk 2 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/book/CreateBookDto.java]: ++, updated. thank you!

File: src/main/java/org/example/intro/controller/CategoryController.java (hunk 1 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CategoryController.java]: ++, thank you. updated

File: src/main/java/org/example/intro/controller/CategoryController.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@Tag(name = "Category management", description = "Endpoints for categories management")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CategoryController.java]:
```suggestion
@RequestMapping("/categories")
```

File: src/main/java/org/example/intro/mapper/BookMapper.java [hunk 10]
```diff
@@ -1,14 +1,42 @@
...
+        List<Long> categoryIds = book.getCategories().stream()
+                .map(Category::getId)
+                .toList();
+        bookDto.setCategoryIds(categoryIds);
+    }
 
-    Book toBook(CreateBookDto createDto);
+    @Mapping(target = "categories", ignore = true)
+    Book toEntity(CreateBookDto createDto);
+    @AfterMapping
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/mapper/BookMapper.java]:
```suggestion
    @Mapping(target = "categories", ignore = true)
    Book toEntity(CreateBookDto createDto);
    
    @AfterMapping
```

File: src/main/java/org/example/intro/mapper/BookMapper.java [hunk 11]
```diff
@@ -1,14 +1,42 @@
...
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
-    BookDto toBookDto(Book book);
+    @Mapping(target = "categoryIds", ignore = true)
+    BookDto toDto(Book book);
+    @AfterMapping
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/mapper/BookMapper.java]:
```suggestion
    @Mapping(target = "categoryIds", ignore = true)
    BookDto toDto(Book book);
    
    @AfterMapping
```

File: src/main/java/org/example/intro/mapper/CategoryMapper.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+public interface CategoryMapper {
+    CategoryDto toDto(Category category);
+
+    Category toEntity(CreateCategoryRequestDto requestDto);
+
+    void updateCategoryFromDto(
+            CreateCategoryRequestDto categoryDto,
+            @MappingTarget Category category
+    );
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/mapper/CategoryMapper.java]:
```suggestion
```

File: src/main/java/org/example/intro/model/Book.java (hunk 4 above)
Comment (Elena-Bruyako) [reply src/main/java/org/example/intro/model/Book.java]: not fixed

File: src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java (hunk 7 above)
Comment (Elena-Bruyako) [reply src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java]: not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/intro/mapper/CategoryMapper.java (hunk 12 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/mapper/CategoryMapper.java]: ++, removed space

File: src/main/java/org/example/intro/mapper/BookMapper.java (hunk 11 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/mapper/BookMapper.java]: ++, added space

File: src/main/java/org/example/intro/mapper/BookMapper.java (hunk 10 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/mapper/BookMapper.java]: ++, added space

File: src/main/java/org/example/intro/controller/CategoryController.java (hunk 9 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CategoryController.java]: ++, updated

File: src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java (hunk 7 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/CategoryServiceImpl.java]: ++

File: src/main/java/org/example/intro/model/Book.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/Book.java]: ++

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**